bundles/org.eclipse.ltk.core.refactoring/src/org/eclipse/ltk/core/refactoring/resource/DeleteResourceChange.java

```diff
@@ -120,61 +120,55 @@ public String getName() {
-			}
+		SubMonitor subMonitor= SubMonitor.convert(pm, RefactoringCoreMessages.DeleteResourceChange_deleting, 10);
+
+		IResource resource= getResource();
+		if (resource == null || !resource.exists()) {
+			if (fDeleteContent)
+				return null; // see https://bugs.eclipse.org/343584
+			String message= Messages.format(RefactoringCoreMessages.DeleteResourceChange_error_resource_not_exists, BasicElementLabels.getPathLabel(fResourcePath.makeRelative(), false));
+			throw new CoreException(new Status(IStatus.ERROR, RefactoringCorePlugin.getPluginId(), message));
+		}
 
-			// make sure all files inside the resource are saved so restoring works
-			if (resource.isAccessible()) {
-				resource.accept((IResourceVisitor) curr -> {
-					try {
-						if (curr instanceof IFile) {
-							// progress is covered outside.
-							saveFileIfNeeded((IFile) curr, new NullProgressMonitor());
-						}
-					} catch (CoreException e) {
-						// ignore
+		// make sure all files inside the resource are saved so restoring works
+		if (resource.isAccessible()) {
+			resource.accept((IResourceVisitor) curr -> {
+				try {
+					if (curr instanceof IFile) {
+						// progress is covered outside.
+						saveFileIfNeeded((IFile) curr, new NullProgressMonitor());
 					}
-					return true;
-				}, IResource.DEPTH_INFINITE, false);
-			}
+				} catch (CoreException e) {
+					// ignore
+				}
+				return true;
+			}, IResource.DEPTH_INFINITE, false);
+		}
 
-			ResourceUndoState desc= ResourceUndoState.fromResource(resource);
-			if (resource instanceof IProject) {
-				((IProject) resource).delete(fDeleteContent, fForceOutOfSync, new SubProgressMonitor(pm, 10));
+		ResourceUndoState desc= ResourceUndoState.fromResource(resource);
+		if (resource instanceof IProject) {
+			((IProject) resource).delete(fDeleteContent, fForceOutOfSync,subMonitor.newChild(10));
+		} else {
+			int updateFlags;
+			if (fForceOutOfSync) {
+				updateFlags= IResource.KEEP_HISTORY | IResource.FORCE;
 			} else {
-				int updateFlags;
-				if (fForceOutOfSync) {
-					updateFlags= IResource.KEEP_HISTORY | IResource.FORCE;
-				} else {
-					updateFlags= IResource.KEEP_HISTORY;
-				}
-				resource.delete(updateFlags, new SubProgressMonitor(pm, 5));
-				desc.recordStateFromHistory(resource, new SubProgressMonitor(pm, 5));
+				updateFlags= IResource.KEEP_HISTORY;
 			}
-			return new UndoDeleteResourceChange(desc);
-		} finally {
-			pm.done();
+			resource.delete(updateFlags, subMonitor.newChild(5));
+			desc.recordStateFromHistory(resource, subMonitor.newChild(5));
 		}
+		return new UndoDeleteResourceChange(desc);
+
 	}
 
 	private static void saveFileIfNeeded(IFile file, IProgressMonitor pm) throws CoreException {
 		ITextFileBuffer buffer= FileBuffers.getTextFileBufferManager().getTextFileBuffer(file.getFullPath(), LocationKind.IFILE);
 		if (buffer != null && buffer.isDirty() && buffer.isStateValidated() && buffer.isSynchronized()) {
+			SubMonitor subMonitor= SubMonitor.convert(pm, 2);
```

Comment:
convert should be enough, you do not need beginiTask (next line) anymore

Reply:
Done